invoice/src/data.rs

```diff
@@ -0,0 +1,112 @@
+use std::convert::Infallible;
+use std::str::FromStr;
+
+use rgb::{DataState, KnownState, RevealedData};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+use strict_encoding::{StrictDeserialize, StrictSerialize};
+
+use crate::LIB_NAME_RGB_CONTRACT;
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug, Display)]
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate", rename_all = "camelCase")
+)]
+pub enum NonFungible {
+    #[display(inner)]
+    RGB21(Allocation),
+}
+
+impl FromStr for NonFungible {
+    type Err = Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(if let Ok(allocation) = Allocation::from_str(s) {
+            NonFungible::RGB21(allocation)
+        } else {
+            unreachable!()
+        })
+    }
+}
```

Comment:
Since all this does is parse the allocation, it should be fine to use the same error.

```suggestion
impl FromStr for NonFungible {
    type Err = AllocationParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let allocation = Allocation::from_str(s)?;
        Ok(NonFungible::RGB21(allocation))
    }
}
```